scripts/staging/fm-regression.dml

```diff
@@ -0,0 +1,172 @@
+
+    # regularization
+    lambd = 5e-04
+
+    # Optimize
+    print("Starting optimization")
+    batch_size = 64
+    epochs = 10
+    iters = ceil(N / batch_size)
+
+    for (e in 1:epochs) {
+      for (i in 1:iters) {
+        # Get the next batch
+        beg = ((i-1) * batch_size) %% N + 1
+        end = min(N, beg + batch_size - 1)
+        X_batch = X[beg:end,]
+        y_batch = y[beg:end,]
+
+        # 3.Send inputs through fm::forward
+        out = fm::forward(X_batch, w0, W, V);
+
+        # 4.Send the above result through log_loss::forward
+        sfy = sigmoid::forward(out);
+
+        # 5.compute gradients from a loss l2_loss::backward
+        dy = l2_loss::backward(sfy, out);
+
+        # Compute loss & accuracy for training & validation data every 100 iterations.
+        if (i %% 100 == 0) {
+          # Compute training loss & accuracy
+          loss_data = cross_entropy_loss::forward(sfy, y_batch)
+          loss_reg_w0 = l2_reg::forward(w0, lambda)
+          loss_reg_W  = l2_reg::forward(W , lambda)
+          loss_reg_V  = l2_reg::forward(V , lambda)
+          loss = loss_data + loss_reg_w0 + loss_reg_W + loss_reg_V
+          accuracy = mean(rowIndexMax(sfy) == rowIndexMax(y_batch))
+
+
+          # Compute validation loss & accuracy
+          probs_val = predict(X_val, w0, W, V)
+          loss_val  = cross_entropy_loss::forward(probs_val, y_val)
+          accuracy_val = mean(rowIndexMax(probs_val) == rowIndexMax(y_val))
+
+          # Output results
+          print("Epoch: " + e + ", Iter: " + i + ", Train Loss: " + loss + ", Train Accuracy: "
+                 + accuracy + ", Val Loss: " + loss_val + ", Val Accuracy: " + accuracy_val)
+        }
+
+        # 6.Send the above result through fm::backward
+        [dw0, dW, dV] = fm::backward(dy, X, w0, W, V);
+
+        # 6.Call adam::update for all parameters
+        w0 = adam::update(w0, dw0, lr, beta1, beta2, epsilon, t, mw0, vw0);
+        W  = adam::update(W, dW, lr, beta1, beta2, epsilon, t, mW, vW );
+        V  = adam::update(V, dV, lr, beta1, beta2, epsilon, t, mV, vV );
+
+      }
+    }
+}
+
+predict = function(matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] out) {
+  /*
+   * Computes the predictions for the given inputs.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - w0, W, V : trained model parameters.
+   *
+   * Outputs:
+   *  - out : target vector, y.
+   */
+
+    # 1.initialize fm core
+    [w0, W, V] = fm::init(d, k);
+
+    # 2.Send inputs through fm::forward
+    out = fm::forward(X, w0, W, V);
+
+    # 3.Send the above result through log_loss::forward
+    # loss = log_loss::forward(out);
+
+}
+
+eval = function(matrix[double] probs, matrix[double] y)
+    return (double loss) {
+   /*
+    * Computes loss and accuracy.
+    */
+
+    # just compute loss with log_loss::forward(probs, y)
+    loss = log_loss::forward(probs, y)
```

Comment:
Replace this with the L2 loss, and add a measure of accuracy, such as the root mean square error (RMSE).